File: compiler/src/lexer.rs
use crate::error::LexerError;
use crate::token::Token;
use phf::phf_map;

static KWDS: phf::Map<&'static str, Token> = phf_map! {
    "mut" => Token::Mut,
    "if" => Token::If,
    "for" => Token::For,
    "con" => Token::Con,
    "match" => Token::Match,
    "true" => Token::True,
    "false" => Token::False,
    "break" => Token::Break,
    "continue" => Token::Continue,
    "defer" => Token::Defer,
    "enum" => Token::Enum,
    "struct" => Token::Struct,
    "pub" => Token::Pub,
    "void" => Token::Void,
    "return" => Token::Return,
};

pub struct Lexer {
    l: usize,
    c: usize,
    pub s: String,
    curr: usize,
    pub tok: Result<Option<Token>, LexerError>,
}

impl Lexer {
    pub fn new(source: String) -> Self {
        Self {
            l: 1,
            c: 1,
            s: source,
            curr: 0,
            tok: Ok(None),
        }
    }
    fn ch(&self) -> char {
        if let Some(c) = self.s.chars().nth(self.curr) {
            c
        } else {
            '\0'
        }
    }
    pub fn next(&mut self) {
        if let Ok(Some(ref t)) = self.tok {
            if matches!(t, Token::Eof) {
                return;
            }
        }
        'redo: loop {
            if self.is_end() {
                self.tok = Ok(Some(Token::Eof));
                return;
            }
            loop {
                if !matches!(self.ch(), ' ' | '\t' | '\n' | '\r') {
                    break;
                }
                self.advance();
            }
            let c = self.ch();
            if c.is_alphabetic() || c == '_' {
                self.ident();
                return;
            }
            match c {
                '\0' => self.tok = Ok(Some(Token::Eof)),
                '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => self.number(),
                '"' => self.string(),
                '`' => self.raw_string(),
                '\'' => self.char(),
                '(' => {
                    self.advance();
                    self.tok = Ok(Some(Token::LParen));
                }
                ')' => {
                    self.advance();
                    self.tok = Ok(Some(Token::RParen));
                }
                '[' => {
                    self.advance();
                    self.tok = Ok(Some(Token::LBrack));
                }
                ']' => {
                    self.advance();
                    self.tok = Ok(Some(Token::RBrack));
                }
                '{' => {
                    self.advance();
                    self.tok = Ok(Some(Token::LBrace));
                }
                '}' => {
                    self.advance();
                    self.tok = Ok(Some(Token::RBrace));
                }
                ',' => {
                    self.advance();
                    self.tok = Ok(Some(Token::Comma));
                }
                ';' => {
                    self.advance();
                    self.tok = Ok(Some(Token::Semicolon));
                }
                ':' => {
                    self.advance();
                    self.tok = Ok(Some(Token::Colon));
                }
                '.' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '.' => Some(Token::DotDot),
                        _ => Some(Token::Dot),
                    });
                    if matches!(self.ch(), '.') {
                        self.advance();
                    }
                }
                '+' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::AddAssign),
                        '+' => Some(Token::AddAdd),
                        _ => Some(Token::Add),
                    });
                    if matches!(self.ch(), '=' | '+') {
                        self.advance();
                    }
                }
                '-' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::SubAssign),
                        '-' => Some(Token::SubSub),
                        _ => Some(Token::Sub),
                    });
                    if matches!(self.ch(), '=' | '-') {
                        self.advance();
                    }
                }
                '*' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::MulAssign),
                        _ => Some(Token::Mul),
                    });
                    if matches!(self.ch(), '=' | '*') {
                        self.advance();
                    }
                }
                '%' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::ModAssign),
                        _ => Some(Token::Mod),
                    });
                    if matches!(self.ch(), '=') {
                        self.advance();
                    }
                }
                '/' => {
                    self.advance();
                    match self.ch() {
                        '=' => self.tok = Ok(Some(Token::DivAssign)),
                        '/' => {
                            self.comment();
                            continue 'redo;
                        }
                        '*' => {
                            self.multi_comment();
                            continue 'redo;
                        }
                        _ => self.tok = Ok(Some(Token::Div)),
                    }
                    if matches!(self.ch(), '=') {
                        self.advance();
                    }
                }
                '^' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::XorAssign),
                        _ => Some(Token::Xor),
                    });
                    if matches!(self.ch(), '=') {
                        self.advance();
                    }
                }
                '<' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::LesserEqual),
                        '<' => Some(Token::Lesser),
                        _ => Some(Token::RightShift),
                    });
                    if matches!(self.ch(), '=' | '<') {
                        self.advance();
                    }
                }
                '>' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::GreaterEqual),
                        '>' => Some(Token::Greater),
                        _ => Some(Token::LeftShift),
                    });
                    if matches!(self.ch(), '=' | '>') {
                        self.advance();
                    }
                }
                '&' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::AndAssign),
                        '&' => Some(Token::AndAnd),
                        _ => Some(Token::And),
                    });
                    if matches!(self.ch(), '=' | '&') {
                        self.advance();
                    }
                }
                '|' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::OrAssign),
                        '|' => Some(Token::OrOr),
                        _ => Some(Token::Or),
                    });
                    if matches!(self.ch(), '=' | '|') {
                        self.advance();
                    }
                }
                '=' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::EqualEqual),
                        _ => Some(Token::Equal),
                    });
                    if matches!(self.ch(), '=') {
                        self.advance();
                    }
                }
                '!' => {
                    self.advance();
                    self.tok = Ok(match self.ch() {
                        '=' => Some(Token::BangEqual),
                        _ => Some(Token::Bang),
                    });
                    if matches!(self.ch(), '=') {
                        self.advance();
                    }
                }
                _ => {}
            }
            break 'redo;
        }
    }
    fn section(&self, start: usize) -> String {
        self.s[start..self.curr].to_owned()
    }
    fn ident(&mut self) {
        let start = self.curr;
        let mut c = self.ch();
        while c.is_ascii_alphanumeric() || c == '_' {
            self.advance();
            c = self.ch();
        }

        let v = self.section(start);
        self.tok = if KWDS.contains_key(v.as_str()) {
            let v: Token = KWDS.get(v.as_str()).unwrap().clone();
            Ok(Some(v))
        } else {
            Ok(Some(Token::Ident(v)))
        }
    }
    fn number(&mut self) {
        let mut floating = false;
        let start = self.curr;
        let mut c = self.ch();
        while c.is_ascii_digit() {
            self.advance();
            c = self.ch();
        }
        if c == '.' {
            floating = true;
            self.advance();
            while c.is_ascii_digit() {
                self.advance();
                c = self.ch();
            }
        }
        let v = self.section(start);
        self.tok = if floating {
            Ok(Some(Token::Float(v.parse::<f64>().unwrap())))
        } else {
            Ok(Some(Token::Int(v.parse::<u128>().unwrap())))
        }
    }
    fn string(&mut self) {
        self.advance();
        let start = self.curr;

        loop {
            match self.ch() {
                '"' => {
                    self.advance();
                    self.tok = Ok(Some(Token::String(self.section(start))));
                    break;
                }
                '\\' => {
                    if !self.escape('"') {
                        self.tok = Err(LexerError::InvalidEscape(self.l, self.c));
                    }
                    continue;
                }
                '\n' => {
                    self.tok = Err(LexerError::NewLineInString(self.l, self.c));
                    break;
                }
                _ => {
                    if self.is_end() {
                        self.tok = Err(LexerError::UnterminatedString(self.l, start));
                        break;
                    }
                }
            }
            self.advance();
        }
    }
    fn raw_string(&mut self) {
        self.advance();
        let start = self.curr;

        loop {
            let c = self.ch();

            if c == '`' {
                self.advance();
                self.tok = Ok(Some(Token::String(self.section(start))));
                break;
            }
            if self.is_end() {
                self.tok = Err(LexerError::UnterminatedString(self.l, start));
                break;
            }
            self.advance();
        }
    }
    fn char(&mut self) {
        let mut n = 0;
        let start = self.curr;
        loop {
            let c = self.ch();

            if c == '\'' {
                if n > 1 {
                    self.tok = Err(LexerError::MoreThanOneCharacterInChar(self.l, self.curr))
                }
                self.advance();
                break;
            }
            if c == '\\' {
                self.advance();
                if !self.escape('\'') {
                    self.tok = Err(LexerError::InvalidEscape(self.l, self.curr));
                }
            }
            if self.is_end() {
                self.tok = Err(LexerError::UnterminatedChar(self.l, start));
                break;
            }
            self.advance();
            n += 1;
        }
    }
    fn comment(&mut self) {
        loop {
            if self.ch() == '\n' || self.is_end() {
                break;
            }
            self.advance();
        }
    }
    fn is_end(&mut self) -> bool {
        self.curr >= self.s.len()
    }
    fn multi_comment(&mut self) {
        loop {
            if self.ch() == '*' {
                self.advance();
                if self.ch() == '/' {
                    self.advance();
                    break;
                }
            }
            if self.is_end() {
                break;
            }
            self.advance();
        }
    }
    fn escape(&mut self, c: char) -> bool {
        matches!(self.ch(), 'a' | 'b' | 'f' | 'n' | 'r' | 't' | 'v' | '\\') || self.ch() == c
    }
    fn advance(&mut self) {
        if self.ch() == '\n' {
            self.l += 1;
            self.c = 0;
        } else {
            self.c += 1;
        }
        self.curr += 1;
    }
}

File: compiler/src/lexer.zig
const std = @import("std");
const Token = @import("token.zig");
const CompilerError = @import("errors.zig").CompilerError;
const KWDS = std.ComptimeStringMap(Token.TokenType, .{
    .{ "mut", .Mut },
    .{ "if", .If },
    .{ "for", .For },
    .{ "con", .Con },
    .{ "match", .Match },
    .{ "true", .True },
    .{ "false", .False },
    .{ "break", .Break },
    .{ "continue", .Continue },
    .{ "defer", .Defer },
    .{ "enum", .Enum },
    .{ "struct", .Struct },
    .{ "pub", .Pub },
    .{ "void", .Void },
    .{ "return", .Return },
    .{ "from", .From },
    .{ "use", .Use },
    .{ "type", .Type },
});

pub const Lexer = struct {
    l: usize,
    c: usize,
    s: []u8,
    curr: usize,
    tok: Token.Token,
    err: ?CompilerError,

    // fns
    pub fn init(s: []u8) Lexer {
        return Lexer{
            .l = 1,
            .c = 1,
            .s = s,
            .curr = 0,
            .tok = .{ .type = .None, .lit = "" },
            .err = null,
        };
    }
    pub fn ch(s: *Lexer) u8 {
        if (s.curr >= s.s.len) {
            return '\\';
        } else {
            return s.s[s.curr];
        }
    }
    pub fn next(s: *Lexer) void {
        if (s.tok.type == .Eof) {
            return;
        }
        redo: while (true) {
            if (s.is_end()) {
                s.tok.type = .Eof;
                s.tok.lit = "";
            }
            while (switch (s.ch()) {
                ' ', '\t', '\n', '\r' => true,
                else => false,
            }) {
                s.advance();
            }
            if (s.is_end()) {
                s.tok.type = .Eof;
                s.tok.lit = "";
                return;
            }
            switch (s.ch()) {
                'a'...'z', 'A'...'Z', '_' => s.ident(),
                '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' => s.number(),
                '"' => s.string(),
                '`' => s.raw_string(),
                '\'' => s.char(),
                '(' => {
                    s.advance();
                    s.tok.type = .LParen;
                    s.tok.lit = "";
                },
                ')' => {
                    s.advance();
                    s.tok.type = .RParen;
                    s.tok.lit = "";
                },
                '[' => {
                    s.advance();
                    s.tok.type = .LBrack;
                    s.tok.lit = "";
                },
                ']' => {
                    s.advance();
                    s.tok.type = .RBrack;
                    s.tok.lit = "";
                },
                '{' => {
                    s.advance();
                    s.tok.type = .LBrace;
                    s.tok.lit = "";
                },
                '}' => {
                    s.advance();
                    s.tok.type = .RBrace;
                    s.tok.lit = "";
                },
                ',' => {
                    s.advance();
                    s.tok.type = .Comma;
                    s.tok.lit = "";
                },
                ';' => {
                    s.advance();
                    s.tok.type = .Semicolon;
                    s.tok.lit = "";
                },
                ':' => {
                    s.advance();
                    s.tok.type = .Colon;
                    s.tok.lit = "";
                },
                '.' => {
                    s.advance();
                    switch (s.ch()) {
                        '.' => {
                            s.tok.type = .DotDot;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Dot;
                            s.tok.lit = "";
                        },
                    }
                },
                '+' => {
                    s.advance();
                    switch (s.ch()) {
                        '+' => {
                            s.tok.type = .AddAdd;
                            s.tok.lit = "";
                        },
                        '=' => {
                            s.tok.type = .AddAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Add;
                            s.tok.lit = "";
                        },
                    }
                },
                '-' => {
                    s.advance();
                    switch (s.ch()) {
                        '-' => {
                            s.tok.type = .SubSub;
                            s.tok.lit = "";
                        },
                        '=' => {
                            s.tok.type = .SubAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Sub;
                            s.tok.lit = "";
                        },
                    }
                },
                '*' => {
                    s.advance();
                    switch (s.ch()) {
                        '=' => {
                            s.tok.type = .MulAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Mul;
                            s.tok.lit = "";
                        },
                    }
                },
                '%' => {
                    s.advance();
                    switch (s.ch()) {
                        '=' => {
                            s.tok.type = .ModAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Mod;
                            s.tok.lit = "";
                        },
                    }
                },
                '/' => {
                    s.advance();
                    switch (s.ch()) {
                        '/' => {
                            s.comment();
                            continue :redo;
                        },
                        '*' => {
                            s.multi_comment();
                            continue :redo;
                        },
                        '=' => {
                            s.tok.type = .DivAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Div;
                            s.tok.lit = "";
                        },
                    }
                    s.advance();
                },
                '^' => {
                    s.advance();
                    switch (s.ch()) {
                        '=' => {
                            s.tok.type = .XorAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Xor;
                            s.tok.lit = "";
                        },
                    }
                    s.advance();
                },
                '<' => {
                    s.advance();
                    switch (s.ch()) {
                        '<' => {
                            s.tok.type = .LeftShift;
                            s.tok.lit = "";
                        },
                        '=' => {
                            s.tok.type = .LesserEqual;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Lesser;
                            s.tok.lit = "";
                        },
                    }
                },
                '>' => {
                    s.advance();
                    switch (s.ch()) {
                        '>' => {
                            s.tok.type = .RightShift;
                            s.tok.lit = "";
                        },
                        '=' => {
                            s.tok.type = .GreaterEqual;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Greater;
                            s.tok.lit = "";
                        },
                    }
                },
                '&' => {
                    s.advance();
                    switch (s.ch()) {
                        '&' => {
                            s.tok.type = .AndAnd;
                            s.tok.lit = "";
                        },
                        '=' => {
                            s.tok.type = .AndAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .And;
                            s.tok.lit = "";
                        },
                    }
                },
                '|' => {
                    s.advance();
                    switch (s.ch()) {
                        '|' => {
                            s.tok.type = .OrOr;
                            s.tok.lit = "";
                        },
                        '=' => {
                            s.tok.type = .OrAssign;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Or;
                            s.tok.lit = "";
                        },
                    }
                },
                '=' => {
                    s.advance();
                    switch (s.ch()) {
                        '=' => {
                            s.tok.type = .EqualEqual;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Equal;
                            s.tok.lit = "";
                        },
                    }
                },
                '!' => {
                    s.advance();
                    switch (s.ch()) {
                        '=' => {
                            s.tok.type = .BangEqual;
                            s.tok.lit = "";
                        },
                        else => {
                            s.tok.type = .Bang;
                            s.tok.lit = "";
                        },
                    }
                },
                else => {
                    s.err = CompilerError.UndefinedSymbol;
                },
            }
            break :redo;
        }
    }
    pub fn section(s: *Lexer, start: usize) []u8 {
        return s.s[start..s.curr];
    }
    pub fn ident(s: *Lexer) void {
        const start = s.curr;
        while (switch (s.ch()) {
            'a'...'z', 'A'...'Z', '_', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' => true,
            else => false,
        }) {
            s.advance();
        }
        const v = s.section(start);
        if (KWDS.has(v)) {
            s.tok.type = KWDS.get(v).?;
            s.tok.lit = "";
        } else {
            s.tok.type = .Ident;
            s.tok.lit = v;
        }
    }
    pub fn number(s: *Lexer) void {
        var floating = false;
        const start = s.curr;
        while (switch (s.ch()) {
            '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' => true,
            else => false,
        }) {
            s.advance();
        }
        if (s.ch() == '.') {
            floating = true;
            s.advance();
            while (switch (s.ch()) {
                '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' => true,
                else => false,
            }) {
                s.advance();
            }
        }
        const v = s.section(start);
        if (floating) {
            s.tok.type = .Float;
        } else {
            s.tok.type = .Int;
        }
        s.tok.lit = v;
    }
    pub fn string(s: *Lexer) void {
        s.advance();
        const start = s.curr;
        const start_col = s.c;
        const start_line = s.l;

        while (true) {
            switch (s.ch()) {
                '"' => {
                    s.tok.type = .String;
                    s.tok.lit = s.section(start);
                    s.advance();
                    break;
                },
                '\\' => {
                    if (!s.escape('"')) {
                        s.err = CompilerError.InvalidEscape;
                        s.l = start_line;
                        s.c = start_col;
                        break;
                    }
                },
                '\n' => {
                    s.err = CompilerError.NewlineInSingleLineString;
                    s.l = start_line;
                    s.c = start_col;
                    break;
                },
                else => {
                    if (s.is_end()) {
                        s.err = CompilerError.UnterminatedChar;
                        s.l = start_line;
                        s.c = start_col;
                        break;
                    }
                },
            }
            s.advance();
        }
    }
    pub fn raw_string(s: *Lexer) void {
        s.advance();
        const start = s.curr;
        const start_col = s.c;
        const start_line = s.l;
        while (true) {
            switch (s.ch()) {
                '`' => {
                    s.tok.type = .String;
                    s.tok.lit = s.section(start);
                    s.advance();
                },
                else => {
                    if (s.is_end()) {
                        s.err.? = CompilerError.UnterminatedChar;
                        s.l = start_line;
                        s.c = start_col;
                        break;
                    }
                },
            }
            s.advance();
        }
    }
    pub fn char(s: *Lexer) void {
        var n: usize = 0;
        const start = s.curr;
        const start_col = s.c;
        const start_line = s.l;
        while (true) {
            switch (s.ch()) {
                '\'' => {
                    if (n > 1) {
                        s.err.? = CompilerError.CharLiteralMoreThanOne;
                        s.l = start_line;
                        s.c = start_col;
                    }
                    s.tok.type = .Char;
                    s.tok.lit = s.section(start);
                    s.advance();
                    break;
                },
                '\\' => {
                    s.advance();
                    if (!s.escape('\'')) {
                        s.err.? = CompilerError.InvalidEscape;
                        s.l = start_line;
                        s.c = start_col;
                        break;
                    }
                },
                else => {
                    if (s.is_end()) {
                        s.err.? = CompilerError.UnterminatedChar;
                        s.l = start_line;
                        s.c = start_col;
                        break;
                    }
                },
            }
            s.advance();
            n += 1;
        }
    }
    pub fn comment(s: *Lexer) void {
        while (true) {
            if (s.ch() == '\n' or s.is_end()) {
                break;
            }
            s.advance();
        }
    }
    pub fn multi_comment(s: *Lexer) void {
        while (true) {
            if (s.ch() == '*') {
                s.advance();
                if (s.ch() == '/') {
                    s.advance();
                    break;
                }
            }
            if (s.is_end()) {
                break;
            }
            s.advance();
        }
    }
    pub fn is_end(s: *Lexer) bool {
        return s.curr >= s.s.len;
    }
    pub fn escape(s: *Lexer, c: u8) bool {
        if (s.ch() == c) {
            return true;
        }
        return switch (s.ch()) {
            'a', 'b', 'f', 'n', 'r', 't', 'v', '\\' => true,
            else => false,
        };
    }
    pub fn advance(s: *Lexer) void {
        if (s.ch() == '\n') {
            s.l += 1;
            s.c = 0;
        } else {
            s.c += 1;
        }
        s.curr += 1;
    }
};
